ca/ca_test.go

```diff
@@ -0,0 +1,189 @@
+package ca
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/bwesterb/mtc"
+)
+
+// testCA is a Merkle Tree CA that runs in memory.
+type testCA struct {
+	*Handle
+	path string
+}
+
+// newTestCA creates a CA in a temporary directory. The caller should call
+// `Close()` in order to delete the directory once the test is finished.
+func newTestCA(t *testing.T) *testCA {
+	path, err := os.MkdirTemp("", "test-mtca-*")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	issuer := mtc.RelativeOID{}
+	err = issuer.UnmarshalText([]byte("1.2.3.4"))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	handle, err := New(path, NewOpts{
+		Issuer:        issuer,
+		ServerPrefix:  "ca.example.com",
+		BatchDuration: time.Duration(0),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return &testCA{
+		handle,
+		path,
+	}
+}
+
+func (ca *testCA) Close() {
+	ca.Handle.Close()
+	if err := os.RemoveAll(ca.path); err != nil {
+		panic(err)
+	}
+}
+
+func createVerifyTest(t *testing.T) (*mtc.BikeshedCertificate,
+	*mtc.SignedValidityWindow, mtc.CAParams) {
+	ca := newTestCA(t)
+	defer ca.Close()
+
+	sk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	subject, err := mtc.NewTLSSubject(
+		mtc.TLSECDSAWithP256AndSHA256,
+		sk.Public(),
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	assertion := mtc.Assertion{
+		Subject: subject,
+		Claims: mtc.Claims{
+			DNS: []string{"example.org"},
+		},
+	}
+
+	notAfter := time.Now().Add(time.Hour * 24 * 7)
+
+	assertionRequest := &mtc.AssertionRequest{
+		Assertion: assertion,
+		NotAfter:  notAfter,
+	}
+	err = assertionRequest.Check()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = ca.Queue(*assertionRequest)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	err = ca.Issue()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cert, err := ca.CertificateFor(assertionRequest.Assertion)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	signedValidityWindow, err := ca.SignedValidityWindowForBatch(0)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return cert, signedValidityWindow, ca.Params()
+}
+
+func TestVerifyOk(t *testing.T) {
+	cert, signedValidityWindow, ca := createVerifyTest(t)
+	err := mtc.Verify(cert, mtc.VerifyOptions{
+		CA:             &ca,
+		ValidityWindow: &signedValidityWindow.ValidityWindow,
+		CurrentTime:    time.Now(),
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+func TestVerifyErrorExpired(t *testing.T) {
+	cert, signedValidityWindow, ca := createVerifyTest(t)
+	err := mtc.Verify(cert, mtc.VerifyOptions{
+		CA:             &ca,
+		ValidityWindow: &signedValidityWindow.ValidityWindow,
+		// Set the current time to certificate not after.
+		CurrentTime: cert.Proof.NotAfter(),
```

Comment:
The NotAfter is inclusive, right? I wouldn't expect this test to fail.

Reply:
I understand it to be exclusive: the certificate is not valid after this time, therefore if the current time == notAfter, then the certificate is not valid as of the current time?

Reply:
It's not valid after this time, but it is valid **at** this time.

Reply:
Hmmm ... ok. Will do.

Reply:
In X.509 certs, it says notAfter is inclusive: https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.5

Reply:
thanks :)